### DISCUSSION TITLE
Improve action detection in exec action check?

### DISCUSSION THREAD

**Discussion by dsngo** [statelyai/xstate#1873] [General]
Recently, I have ran to a problem that if I create a function like this:
```js
const {log} = actions;
export const debugStore = (notifier) =>
  // this will log if the environment is `test`, other I would like it to be removed, I have tried with
  test && log((...args) => args, `STORE: ${notifier}`);
  test ? log((...args) => args, `STORE: ${notifier}`) : undefined;
  // and both failed.
  // ony what would work is replace that with {} or add filter Boolean to actions array (which I find it a bit cumbersome, since I want to keep the machine as pure xstate as possible)
  test ? log((...args) => args, `STORE: ${notifier}`) :{}; 
```
It turns out that the actions check for `action.type`, which is undefined or boolean on the top cases. This would be nice if the actions key in the machine can handle these, so it would be easier to attach log action creator into the machine similar to this example without causing any error.
```js
{
  entry: debugStore("enter"),
  // ..
  actions: [debugStore("exec"), "someOperation"],
  // ..
  exit: [debugStore("exit")m "someCleanup"]
}
```

Edit: added sample [CodeSandbox](https://codesandbox.io/s/xstate-example-template-vanilla-forked-df57l?file=/src/index.js).
Edit 2: I find some strange behavior as well, `actions` does not seem to get called in the example above?

**Comment by davidkpiano** [comment]
✏️ Please create a [CodeSandbox (XState template)](https://codesandbox.io/s/xstate-example-template-qcwed) reproduction of this. Thanks!

**Reply by dsngo** [reply]
Thanks, I just added the sample.

**Comment by Andarist** [comment]
You have an error in your machine. Instead of this:
```js
on: { TOGGLE: "active", actions: "someAction2" }
```
you should have this:
```js
on: { TOGGLE: { target: "active", actions: "someAction2" } }
```

What you have there right now would mean that you have a transition for an event named "actions" that targets a state "someAction2".

You have also used `actions` property on at least 2 states - which is incorrect, only transitions can have `actions` property.

**Reply by dsngo** [reply]
That was my bad, thank you @Andarist, I guess I need to switch to ts to really prevent that kind of silly mistake.

What is your take on the issue when I change the `test` variable on the sample above to `true`. The machine failed. 
```js
- const test = false;
+ const test = true;
```
![image](https://user-images.githubusercontent.com/27849964/105658443-0a715800-5ef9-11eb-8969-391099575eb4.png)
 and the case where if I uncomment the line
 ```js
 // ..
 - test && actions.log((...args) => args, `STORE: ${notifier}`); 
 + test ? actions.log((...args) => args, `STORE: ${notifier}`) : undefined; 
 ```
![image](https://user-images.githubusercontent.com/27849964/105658521-32f95200-5ef9-11eb-84dc-d66e13ac95da.png)
